mission_control/src/mission.cpp

```diff
@@ -36,359 +36,67 @@
-      cur_behavior = getNextAbortBehavior();
-    }
-    while (cur_behavior != NULL);
+bool Mission::loadBehavior(const std::string &missionFullPath){
+//  TODO Check for erros and return -1 if were.
+  _missionFullPath = missionFullPath;
+  if(_missionFullPath!=""){
+    _missionTree = _missionFactory.createTreeFromFile(_missionFullPath);
+    printTreeRecursively( _missionTree.rootNode() );
   }
+  return true;
 }
 
-void Mission::addBehavior(Behavior* behavior) { m_behaviors.push_back(behavior); }
-
-Behavior* Mission::getNextBehavior(bool reset)
-{
-  static bool firsttime = true;
-  if (reset || (firsttime == true))
-  {
-    m_behavior_iterator = m_behaviors.begin();
-    firsttime = false;
-  }
-
-  if (m_behavior_iterator == m_behaviors.end())
-  {
-    return NULL;
-  }
-
-  return *m_behavior_iterator++;
-}
-
-void Mission::addAbortBehavior(Behavior* behavior) { m_aborts.push_back(behavior); }
-
-Behavior* Mission::getNextAbortBehavior(bool reset)
+void Mission::Stop()
 {
-  static bool abortfirsttime = true;
-  if (reset || (abortfirsttime == true))
-  {
-    m_abort_behavior_iterator = m_aborts.begin();
-    abortfirsttime = false;
-  }
-
-  if (m_abort_behavior_iterator == m_aborts.end())
-  {
-    return NULL;
-  }
-
-  return *m_abort_behavior_iterator++;
+  _missionTree.haltTree();
 }
 
-void Mission::ProcessCorrectedPoseData(const pose_estimator::CorrectedData& data)
+int Mission::getMissionStatus()
 {
-  if (m_current_behavior == NULL) return;
-
-  if ((GetState() != MissionState::ABORTING) && (GetState() != MissionState::EXECUTING)) return;
-
-  boost::mutex::scoped_lock callback_lock(m_mutCallbacks);
-  if (m_current_behavior->checkCorrectedData(data)) callBackTmr.stop();
-  callback_lock.unlock();
+  return 0;
 }
 
-void Mission::SetState(MissionState state)
+std::string Mission::getCurrenMissionDescription()
 {
-  boost::mutex::scoped_lock run_lock(m_MissionStateLock);
-  missionState = state;
-  run_lock.unlock();
+  return "Current Mission";
 }
 
-Mission::MissionState Mission::GetState()
+std::string Mission::getCurrentBehavioralName()
 {
-  MissionState temp;
-
-  boost::mutex::scoped_lock run_lock(m_MissionStateLock);
-  temp = missionState;
-  run_lock.unlock();
-
-  return temp;
+  return "Current Behavioral Name";
 }
 
-void Mission::AbortMissionWrapper(ros::NodeHandle nh)
+int Mission::getCurrentBehaviorID()
 {
-  node_handle = nh;
-
-  AbortMission();
+  return 0;
 }
 
-void Mission::ExecuteMission(ros::NodeHandle nh)
+void Mission::ExecuteMission()
 {
-  if ((GetState() == MissionState::EXECUTING) || (GetState() == MissionState::ABORTING))
-  {
-    ROS_WARN("Mission is already executing or aborting");
-    return;
-  }
-
-  SetState(MissionState::EXECUTING);
-  node_handle = nh;
-  // Start the mission processing thread
-  m_threadProcMission = boost::shared_ptr<boost::thread>(
-      new boost::thread(boost::bind(&Mission::processMission, this)));
+  _missionStatus = _missionTree.tickRoot();
```

Comment:
@patapufate how is the execution model going to look like? If it's synchronous you'll need a multi-threaded spinner to handle ROS callbacks. Or push mission execution to another thread, manually.

Reply:
Changed to Asyncronous.